File: src/flavor-go/pkg/psp/operations/operation.go
package operations

import (
	"fmt"
	"io"
)

// Operation constants matching Python/Rust
const (
	// No operation - raw data
	OP_NONE = 0x00

	// Bundle operations (0x01-0x0F)
	OP_TAR = 0x01 // POSIX TAR archive

	// Compression operations (0x10-0x2F)
	OP_GZIP  = 0x10 // GZIP compression
	OP_BZIP2 = 0x13 // BZIP2 compression
	OP_XZ    = 0x16 // XZ/LZMA2 compression
	OP_ZSTD  = 0x1B // Zstandard compression
)

// Operation represents a single transformation operation
type Operation interface {
	// ID returns the operation identifier (e.g., OP_GZIP)
	ID() uint8

	// Name returns the human-readable name
	Name() string

	// Apply applies the operation to input data
	Apply(input []byte) ([]byte, error)

	// ApplyStream applies the operation to a stream
	ApplyStream(input io.Reader, output io.Writer) error

	// Reverse reverses the operation (e.g., decompress for compression)
	Reverse(input []byte) ([]byte, error)

	// ReverseStream reverses the operation on a stream
	ReverseStream(input io.Reader, output io.Writer) error

	// CanReverse returns true if the operation is reversible
	CanReverse() bool

	// EstimateSize estimates the output size given input size
	EstimateSize(inputSize int64) int64
}

// BaseOperation provides common functionality for operations
type BaseOperation struct {
	OpID   uint8
	OpName string
}

func (o *BaseOperation) ID() uint8 {
	return o.OpID
}

func (o *BaseOperation) Name() string {
	return o.OpName
}

func (o *BaseOperation) CanReverse() bool {
	return true // Most operations are reversible
}

func (o *BaseOperation) EstimateSize(inputSize int64) int64 {
	return inputSize // Default: same size
}

// Registry maps operation IDs to implementations
var Registry = make(map[uint8]Operation)

// Register registers an operation implementation
func Register(op Operation) {
	Registry[op.ID()] = op
}

// Get retrieves an operation by ID
func Get(id uint8) (Operation, error) {
	op, ok := Registry[id]
	if !ok {
		return nil, fmt.Errorf("unknown operation: 0x%02x", id)
	}
	return op, nil
}

// GetName returns the name of an operation by ID
func GetName(id uint8) string {
	switch id {
	case OP_NONE:
		return "NONE"
	case OP_TAR:
		return "TAR"
	case OP_GZIP:
		return "GZIP"
	case OP_BZIP2:
		return "BZIP2"
	case OP_XZ:
		return "XZ"
	case OP_ZSTD:
		return "ZSTD"
	default:
		return fmt.Sprintf("UNKNOWN_%02x", id)
	}
}

File: src/flavor-rs/src/psp/operations/operation.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// helpers/flavor-rs/src/psp/operations/operation.rs
// Base operation trait and types

use std::io::{Read, Write};

pub type OperationResult<T> = Result<T, OperationError>;

#[derive(Debug, thiserror::Error)]
pub enum OperationError {
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
    
    #[error("Compression error: {0}")]
    Compression(String),
    
    #[error("Archive error: {0}")]
    Archive(String),
    
    #[error("Operation not reversible")]
    NotReversible,
    
    #[error("Unknown operation: 0x{0:02x}")]
    UnknownOperation(u8),
    
    #[error("Invalid data: {0}")]
    InvalidData(String),
}

/// Trait for all operations
pub trait Operation: Send + Sync {
    /// Returns the operation identifier
    fn id(&self) -> u8;
    
    /// Returns the human-readable name
    fn name(&self) -> &str;
    
    /// Applies the operation to input data
    fn apply(&self, input: &[u8]) -> OperationResult<Vec<u8>>;
    
    /// Applies the operation to a stream
    fn apply_stream(&self, input: &mut dyn Read, output: &mut dyn Write) -> OperationResult<()>;
    
    /// Reverses the operation (e.g., decompress for compression)
    fn reverse(&self, input: &[u8]) -> OperationResult<Vec<u8>>;
    
    /// Reverses the operation on a stream
    fn reverse_stream(&self, input: &mut dyn Read, output: &mut dyn Write) -> OperationResult<()>;
    
    /// Returns true if the operation is reversible
    fn can_reverse(&self) -> bool {
        true
    }
    
    /// Estimates the output size given input size
    fn estimate_size(&self, input_size: u64) -> u64 {
        input_size
    }
}

/// Get operation name by ID
pub fn get_name(id: u8) -> &'static str {
    match id {
        super::OP_NONE => "NONE",
        super::OP_TAR => "TAR",
        super::OP_GZIP => "GZIP",
        super::OP_BZIP2 => "BZIP2",
        super::OP_XZ => "XZ",
        super::OP_ZSTD => "ZSTD",
        _ => "UNKNOWN",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn get_name_returns_correct_names_for_known_ops() {
        assert_eq!(get_name(super::super::OP_NONE), "NONE");
        assert_eq!(get_name(super::super::OP_TAR), "TAR");
        assert_eq!(get_name(super::super::OP_GZIP), "GZIP");
        assert_eq!(get_name(super::super::OP_BZIP2), "BZIP2");
        assert_eq!(get_name(super::super::OP_XZ), "XZ");
        assert_eq!(get_name(super::super::OP_ZSTD), "ZSTD");
    }

    #[test]
    fn get_name_returns_unknown_for_unrecognized_ops() {
        assert_eq!(get_name(0xFF), "UNKNOWN");
        assert_eq!(get_name(0x50), "UNKNOWN");
    }

    #[test]
    fn operation_error_display_formats_correctly() {
        let io_err = OperationError::Io(std::io::Error::new(
            std::io::ErrorKind::Other,
            "disk full",
        ));
        assert!(io_err.to_string().contains("disk full"));

        let comp_err = OperationError::Compression("bad data".into());
        assert!(comp_err.to_string().contains("bad data"));

        let arch_err = OperationError::Archive("corrupt".into());
        assert!(arch_err.to_string().contains("corrupt"));

        let not_rev = OperationError::NotReversible;
        assert!(not_rev.to_string().contains("not reversible"));

        let unknown = OperationError::UnknownOperation(0xAB);
        assert!(unknown.to_string().contains("0xab"));

        let invalid = OperationError::InvalidData("bad".into());
        assert!(invalid.to_string().contains("bad"));
    }
}